extmod/libmetal/metal/system/micropython/alloc.h

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2016, Xilinx Inc. and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/*
+ * @file    micropython/alloc.h
+ * @brief   MicroPython libmetal memory allocation definitions.
+ */
+
+#ifndef __METAL_ALLOC__H__
+#error "Include metal/alloc.h instead of metal/micropython/alloc.h"
+#endif
+
+#ifndef __METAL_MICROPYTHON_ALLOC__H__
```

Comment:
I noticed you've changed these from the generic versions. My intention was to have as few diffs as possible, and to annotate each line that was changed with a `// Micropython...` comment.

I talked to @dpgeorge about this, and we think there's a simpler solution that means we don't have to copy any boilerplate, and in particular avoids duplicating or having to introduce a new license header.

1. Replace each of `alloc.h`, `log.h`, `sleep.h`, `sys.h` with a file that just contains a single line:

```c
#include <metal/micropython.h>
```

3. Then `extmod/libmetal/metal/micropython.h` contains the functional bits of each of these four files, with none of `#ifdefs`, `extern "C"`, copyright headers, etc.

```c
// MicroPython copyright header...

#ifndef MICROPY_INCLUDED_METAL_MICROPYTHON_H
#define MICROPY_INCLUDED_METAL_MICROPYTHON_H

#include <stdlib.h>
#include <metal/config.h>

// generic/alloc.h
static inline void *__metal_allocate_memory(unsigned int size) {
    // MicroPython: changed from "malloc()"
    return m_tracked_calloc(1, size);
}

static inline void __metal_free_memory(void *ptr) {
    // MicroPython: changed from "free()"
    m_tracked_free(ptr);
}

// generic/log.h
#if METAL_LOG_HANDLER_ENABLE
#include "py/mphal.h"
#undef metal_log
#define metal_log(level, ...) mp_printf(&mp_plat_print, __VA_ARGS__)
#endif

// generic/sleep.h
// Must be provided by the port (in mpmetalport.h / mpmetalport.c).
int __metal_sleep_usec(unsigned int usec);

// sys.h
#ifndef METAL_MAX_DEVICE_REGIONS
#define METAL_MAX_DEVICE_REGIONS 1
#endif

struct metal_state {
    struct metal_common_state common;
};

#ifdef METAL_INTERNAL

// generic/template/sys.h
void sys_irq_enable(unsigned int vector);
void sys_irq_disable(unsigned int vector);

// generic/sys.h
void sys_irq_restore_enable(unsigned int flags);
unsigned int sys_irq_save_disable(void);
#endif

#endif MICROPY_INCLUDED_METAL_MICROPYTHON_H
```

Reply:
Done, but we can just add this to `config.h` ?

Reply:
After doing this change, it broke the build because `sys.h` includes `metal/micropython.h` before it defines `struct metal_common_state`. I had to move the struct back to `system/micropython/sys.h`

```
In file included from build-ARDUINO_GIGA/openamp/metal/system/micropython/log.h:1,
                 from build-ARDUINO_GIGA/openamp/metal/log.h:114,
                 from build-ARDUINO_GIGA/openamp/metal/sys.h:18,
                 from ../../extmod/modopenamp.c:34:
build-ARDUINO_GIGA/openamp/metal/micropython.h:63:31: error: field 'common' has incomplete type
   63 |     struct metal_common_state common;
```

Reply:
I moved everything to `config.h`.